internal/controller/promotions/promoqueues.go

```diff
@@ -0,0 +1,156 @@
+type promoQueues struct {
+	// activePromoByStage holds the active promotion for a given stage (if any)
+	activePromoByStage map[types.NamespacedName]string
+	// pendingPromoQueuesByStage holds a priority queue of promotions, per Stage. We allow one
+	// promotion to run at a time, ordered by creationTimestamp.
+	pendingPromoQueuesByStage map[types.NamespacedName]runtime.PriorityQueue
+	// promoQueuesByStageMu protects access to the above maps
+	promoQueuesByStageMu sync.RWMutex
+}
+
+func newPriorityQueue() runtime.PriorityQueue {
+	// We can safely ignore errors here because the only error that can happen
+	// involves initializing the queue with a nil priority function, which we
+	// know we aren't doing.
+	pq, _ := runtime.NewPriorityQueue(func(left, right client.Object) bool {
+		if left.GetCreationTimestamp().Time.Equal(
+			right.GetCreationTimestamp().Time,
+		) {
+			return left.GetName() < right.GetName()
+		}
+		return left.GetCreationTimestamp().Time.
+			Before(right.GetCreationTimestamp().Time)
+	})
+	return pq
+}
+
+// initializeQueues adds the promotion list to relevant priority queues.
+// This is intended to be invoked ONCE and the caller MUST ensure that.
+func (pqs *promoQueues) initializeQueues(ctx context.Context, promos kargoapi.PromotionList) {
+	pqs.promoQueuesByStageMu.Lock()
+	defer pqs.promoQueuesByStageMu.Unlock()
+	logger := logging.LoggerFromContext(ctx)
+	for _, promo := range promos.Items {
+		promo := promo // This is to sidestep implicit memory aliasing in this for loop
+		if promo.Status.Phase.IsTerminal() || promo.Spec == nil {
+			continue
+		}
+		stage := types.NamespacedName{
+			Namespace: promo.Namespace,
+			Name:      promo.Spec.Stage,
+		}
+		pq, ok := pqs.pendingPromoQueuesByStage[stage]
+		if !ok {
+			pq = newPriorityQueue()
+			pqs.pendingPromoQueuesByStage[stage] = pq
+		}
+		if promo.Status.Phase == kargoapi.PromotionPhaseRunning {
+			if pqs.activePromoByStage[stage] == "" {
+				pqs.activePromoByStage[stage] = promo.Name
+			}
+			continue
+		}
+		pq.Push(&promo)
+		logger.WithFields(log.Fields{
+			"promotion": promo.Name,
+			"namespace": promo.Namespace,
+			"stage":     promo.Spec.Stage,
+			"phase":     promo.Status.Phase,
+		}).Debug("pushed Promotion onto Stage-specific Promotion queue")
+	}
+	if logger.Logger.IsLevelEnabled(log.DebugLevel) {
+		for stage, pq := range pqs.pendingPromoQueuesByStage {
+			logger.WithFields(log.Fields{
+				"stage":     stage.Name,
+				"namespace": stage.Namespace,
+				"depth":     pq.Depth(),
+			}).Debug("Stage-specific Promotion queue initialized")
+		}
+	}
+}
+
+// tryActivate tries to mark the given Pending promotion as the active one so it can reconcile.
+// Returns true if the promo is already active or became active as a result of this call.
+// Returns false if it should not reconcile (another promo is active, or next in line).
+func (pqs *promoQueues) tryActivate(ctx context.Context, promo *kargoapi.Promotion) bool {
```

Comment:
Nit: I think "begin" and "conclude" would be more easily understandable verbs on a promotion than activate/deactivate.

Although I suggest renaming those functions, I wouldn't go as far as to suggest renaming variables containing the word "active," as I think the meaning is clear enough in those cases. i.e. Rename functions for clarity, but leave the implementations as they are.

Also, not a hill I'd die on, so do what you will with this suggestion.

Reply:
I don't mind renaming. I chose activate because it's becoming active, but am fine with begin/conclude as well. Will make the change

Reply:
Renamed to `tryBegin` and `conclude`.